### PR TITLE
provider: Add AWSClient.StopContext

### PR DESCRIPTION
**Pull request by bflad**
Add `?w=1` to the PR files URL make it legible.

Reference: #4177 

Provider smoke tests okay in the test framework 😉 

```
make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsAvailabilityZone'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccDataSourceAwsAvailabilityZone -timeout 120m
=== RUN   TestAccDataSourceAwsAvailabilityZone
--- PASS: TestAccDataSourceAwsAvailabilityZone (9.45s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	9.493s
```

### CODE REVIEW

File: aws/provider.go [hunk 1]
```diff
@@ -594,8 +594,11 @@ func Provider() terraform.ResourceProvider {
 			"aws_alb_target_group_attachment": resourceAwsLbTargetGroupAttachment(),
 			"aws_lb_target_group_attachment":  resourceAwsLbTargetGroupAttachment(),
 		},
-		ConfigureFunc: providerConfigure,
 	}
+
+	p.ConfigureFunc = providerConfigure(p)
```
[the remark this answers is not in the document]
Comment (bflad) [reply aws/provider.go]: This is how its implemented in `azurerm`:

https://github.com/terraform-providers/terraform-provider-azurerm/blob/e44d6292405662e4f20bfd21d540a18fa9a2d58f/azurerm/provider.go#L223

File: aws/provider.go [hunk 2]
```diff
@@ -698,86 +701,100 @@ func init() {
...
-		config.SnsEndpoint = endpoints["sns"].(string)
-		config.SqsEndpoint = endpoints["sqs"].(string)
-		config.StsEndpoint = endpoints["sts"].(string)
-	}
+		client.StopContext = p.StopContext()
 
-	if v, ok := d.GetOk("allowed_account_ids"); ok {
-		config.AllowedAccountIds = v.(*schema.Set).List()
-	}
+		// replaces the context between tests
```
[the remark this answers is not in the document]
Comment (bflad) [reply aws/provider.go]: https://godoc.org/github.com/hashicorp/terraform/helper/schema#Provider

```
    // MetaReset is called by TestReset to reset any state stored in the meta
    // interface.  This is especially important if the StopContext is stored by
    // the provider.
    MetaReset func() error
```

Duplicated from `azurerm`:

https://github.com/terraform-providers/terraform-provider-azurerm/blob/e44d6292405662e4f20bfd21d540a18fa9a2d58f/azurerm/provider.go#L279-L283

File: aws/config.go [hunk 3]
```diff
@@ -238,7 +240,7 @@ func (c *AWSClient) IsChinaCloud() bool {
 }
 
 // Client configures and returns a fully initialized AWSClient
-func (c *Config) Client() (interface{}, error) {
+func (c *Config) Client() (*AWSClient, error) {
```
[the remark this answers is not in the document]
Comment (bflad) [reply aws/config.go]: I can switch it back, but I do not see why we need to hide it and it introduces extra type setting. Explicit > Implicit, in my mind anyways. 😄

**Comment by paultyng** [comment]
In your retry func can you do something similar to:

```
select {
  case <-ctx.Done():
    //return non-retryable
  default:
    //keep trying...
}
```

**Comment by bflad** [comment]
Closing this out as it seems it will not be handled this way.

**Pull request closed without merging**